amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/AmqpChannel.java

```diff
@@ -776,38 +359,47 @@ private void setPublishFrame(MessagePublishInfo info, final MessageDestination e
-            connection.sendConnectionClose(NOT_FOUND, "Exchange or queue not found.", channelId);
-            return;
-        }
-        Topic topic = amqpQueue.getIndexTopic();
-        AmqpConsumer amqpConsumer = fetchConsumerMap.computeIfAbsent(queueName, value -> {
+        CompletableFuture<AmqpQueue> amqpQueueCompletableFuture =
+                queueContainer.asyncGetQueue(connection.getNamespaceName(), queue.toString(), false);
+        amqpQueueCompletableFuture.whenComplete((amqpQueue, throwable) -> {
+            if (throwable != null) {
+                log.error("Get Topic error:{}", throwable.getMessage());
+                closeChannel(INTERNAL_ERROR, "Get Topic error: " + throwable.getMessage());
+            } else {
+                if (amqpQueue == null) {
+                    closeChannel(ErrorCodes.NOT_FOUND, "No such queue: " + queueName);
+                    return;
+                } else {
+                    Topic topic = amqpQueue.getTopic();
+                    AmqpConsumer amqpConsumer = fetchConsumerMap.computeIfAbsent(queueName, value -> {
 
-            Subscription subscription = topic.getSubscription(defaultSubscription);
-            AmqpConsumer consumer;
-            try {
-                if (subscription == null) {
-                    subscription = topic.createSubscription(defaultSubscription,
-                        PulsarApi.CommandSubscribe.InitialPosition.Earliest, false).get();
+                        Subscription subscription = topic.getSubscription(defaultSubscription);
+                        AmqpConsumer consumer;
+                        try {
+                            if (subscription == null) {
+                                subscription = topic.createSubscription(defaultSubscription,
+                                        PulsarApi.CommandSubscribe.InitialPosition.Earliest, false).get();
+                            }
+                            consumer = new AmqpPullConsumer(queueContainer, subscription,
+                                    PulsarApi.CommandSubscribe.SubType.Shared,
+                                    topic.getName(), 0, 0, "", 0,
+                                    connection.getServerCnx(), "", null, false,
+                                    PulsarApi.CommandSubscribe.InitialPosition.Latest, null, this,
+                                    "", queueName, noAck);
+                            subscription.addConsumer(consumer);
+                            consumer.handleFlow(DEFAULT_CONSUMER_PERMIT);
+                            return consumer;
+                        } catch (Exception e) {
+                            e.printStackTrace();
+                        }
+                        return null;
+                    });
+                    MessageFetchContext context = MessageFetchContext.get(this, amqpConsumer);
+                    context.handleFetch(noAck);
+                    context.recycle();
```

Comment:
You can't call recycle here. Inner of the context.handleFetch() is an async method for reading entries. If we recycle the context before the read entry callback completed, there will be serious problems.

Reply:
Done.